blockEntity.js

```diff
@@ -3,68 +3,149 @@ const nbt = require('prismarine-nbt')
           }
-        },
+        }
+      }
 
-        set signText (text) {
-          const texts = []
-          if (typeof text === 'string') {
-            // Sign line should look like JSON string of `{"text: "actualText"}`. Since we have plaintext, need to add in this JSON wrapper.
-            texts.push(JSON.stringify(text.split('\n').map((t) => ({ text: t }))))
-          } else if (Array.isArray(text)) {
-            for (const t of text) {
-              if (t.toJSON) { // prismarine-chat
-                texts.push(JSON.stringify(t.toJSON()))
-              } else if (typeof t === 'object') { // normal JS object
-                texts.push(JSON.stringify(t))
-              } else { // plaintext
-                texts.push(JSON.stringify({ text: t }))
+      if (!block.entity) {
+        block.entity = nbt.comp({
+          id: nbt.string(registry.version['>=']('1.11') ? 'minecraft:sign' : 'Sign')
+        })
+      }
+
+      block.entity.value[side].value.messages.value.value = texts
+    }
+
+    return {
+      sign: registry.supportFeature('multiSidedSigns')
+        ? {
+            get blockEntity () {
+              if (!this.entity) return undefined
+
+              const getSideData = (side) => {
+                return {
+                  hasGlowingText: this.entity.value[side].value.has_glowing_text.value === 1,
+                  color: this.entity.value[side].value.color.value,
+                  lines: this.entity.value[side].value.messages.value.value.map(line => new ChatMessage(JSON.parse(line)))
+                }
+              }
+
+              const frontText = getSideData('front_text')
+              return {
+                isWaxed: this.entity.value.is_waxed.value === 1,
+                id: this.entity.value.id || 'minecraft:sign',
+                front: frontText,
+                back: getSideData('back_text'),
+
+                // Backwards compatible API
+                Text1: frontText.lines[0],
+                Text2: frontText.lines[1],
+                Text3: frontText.lines[2],
+                Text4: frontText.lines[3]
               }
+            },
+
+            set signFrontText (text) {
+              setSignText(this, 'front_text', text)
+            },
+
+            get signFrontText () {
+              if (!this.entity) return ''
+              return this.entity.value.front_text.value.messages.value.value.map(text => typeof JSON.parse(text) === 'string' ? JSON.parse(text) : new ChatMessage(JSON.parse(text)).toString()).join('\n')
+            },
+
+            set signBackText (text) {
+              setSignText(this, 'back_text', text)
+            },
+
+            get signBackText () {
+              if (!this.entity) return ''
+              return this.entity.value.back_text.value.messages.value.value.map(text => typeof JSON.parse(text) === 'string' ? JSON.parse(text) : new ChatMessage(JSON.parse(text)).toString()).join('\n')
+            },
```

Comment:
Remove

Reply:
set signBackText is kind of useless but I feel like the getter is kind of useful, to quickly get a plain text version of what's on the sign, although I could definitely change that to use the blockEntity property

Reply:
Using getters/setters is not idiomatic JS (side effects to assignment is not immediately obvious to readers) and is rarely used in most programs aside from places where compatibility is necessary (this is actually why the code is a get/set currently). Like mentioned, it's also not forward compatible in the event changes need to be made like adding parameters.

Reply:
Yeah I'm happy to change it, obviously when I did this I just stuck with how it was done previously, was just confused by you saying remove on those lines specifically. Give me a bit and I'll rework the blockEntity stuff

Reply:
Yeah, the existing .signText can be calling `this.setSignText(front)` and `return this.getSignText()?.[0]` internally. Also, I think it's more future proof to return an array as opposed to front/back to account for future changes.

As for the blockEntity : it's supposed to be the simplified NBT as opposed to the unsimplified NBT inside .entity. Where this matters is when you need to write things back to disk, on the server side. Simplified NBT loses type data that makes it not possible to write back changes. The .blockEntity thing on signs here overrides that functionality because mineflayer internally expects .blockEntity's Text fields to have prismarine-chat ChatMessages instead of stringified JSON. I don't think anything outside mineflayer relies on this behavior, so likely the special .blockEntity handlers here can be completely removed with an update to mineflayer.